### DISCUSSION TITLE
Как правильно сделать отчёт по движению денежных средств aka БДДС в Totum

### DISCUSSION THREAD

**Discussion by gorbarov** [totumonline/totum-ru-issues-and-discussions#358] [Как это сделать?] [answered]
Есть временная таблица Движение денег, в которую из дебеторки (income) и кредиторки (outcome) собирается движение денег.

![image](https://user-images.githubusercontent.com/388620/235044857-0524e6f2-798e-41e1-95ec-9263a35a4e1e.png)
Вот так выглядит, для наглядности.

В хедере есть соответствующие поля, куда собираются списки из двух таблиц.
Код для расходов, для прихода код аналогичный, только ID и сумма идёт без минуса:
```
= : listReplace(list: $rows; action: "id2" = $id2;  action: "id" = $id; action: "move" = $move; action: "type" = $type; action: "desc" = $desc;   value: "val"; )
		id: $#val[id] * -1
		move: $#val[summa_fakt] * -1
		type: 1
		desc: $#val[postavshchik]
		id2: $#val[id]
		
	rows: selectRowList(table: 'outcome'; field: 'id';  field: 'summa_fakt'; field: 'data_fakt'; field: 'postavshchik'; where: 'data_fakt' >= #h_date_from; where: 'summa_fakt' > 0;  order: 'data_fakt' asc)
```

В поле h_insert_result я просто объединяю списки и сортирую их:
```
= : listSort(list: $l1; key: "data_fakt" asc; ) 

l1: listAdd( list: #h_insert_outcome; list: #h_insert_income;)
```

Дальше в колонку insert просто вставляю `= : #h_insert_result[[id]]` айдишники объединённого списка.

Все остальные данные беру из результирующего списка 
```
= : $l[[data_fakt]][0]
~l: listFilter(list: #h_insert_result; key: "item" = #insert; item: "id")
```

**Дальше начинаются затыки, первый из них:**
Чтобы корректно считать остаток за день, мне нужны ID, которые идут по порядку, потому что в одной дате может быть несколько платежей.

Вот так считается остаток сейчас.
![image](https://user-images.githubusercontent.com/388620/235045824-b600621c-a455-4b95-926a-6106de3c7965.png)

Всё работает ровно до того момента, когда не пытаешься выбрать дату раньше, заданной в фильтре. 
Тогда чтобы я не делал, таблица не пересчитывается полностью и получается ситуация типа такой: 
![image](https://user-images.githubusercontent.com/388620/235046168-d1412eaf-a27e-4a00-bd38-d135d78b6256.png)
ID дописываются и привет, начальный остаток прописывается не в той строке и отчёт уже показывает погоду на Марсе.
Если дату двигать вперёд, всё ок.

Нашёл такой [пример](https://ru.docs.totum.online/how-can-I-get-the-previous-value-of-a-field-and-use-it-in-the-action-code), но он работает точно так же как ID

Из-за того, что я сортирую платежи по дате, они идут не в том порядке, как в исходных таблицах, поэтому запрос к исходным таблицам, тоже показывает погоду на марсе.

**В общем это непобедимая оказалась для меня проблема. 
И мне в целом показалось, что я что-то упускаю и есть решение попроще.

Я вижу только одно решение, если бы таблица перезаписывалась целиком, можно ли этого добиться?**

**Comment by erpic-ru** [comment]
Могу предложить 2 варианта:
1. Считать сумму за день и показывать остаток за день. Т.е. складывать приход и расход за день к предыдущему значению. Выглядит это, примерно, так:
```
= : $sel + $sum1 - $sum2
//вычисляем сумму транзакций
sum1:listSum(list: $do1)
sum2:listSum(list: $do2)
///собираем список приходов и расходов за день отдельно
do1:selectList(table: $#ntn; field: 'summa_tranzakcii'; where: 'koshelek' = #koshelek ; where: 'tip' = 1; where: 'timestamp' =#timestamp)
do2:selectList(table: $#ntn; field: 'summa_tranzakcii'; where: 'koshelek' = #koshelek ; where: 'tip' = 2; where: 'timestamp' =#timestamp)
/// берем остаток за предыдущий день
sel: select(table: $#ntn; field: 'balans'; where: 'koshelek' = #koshelek ; where: 'timestamp' < #timestamp; order: 'timestamp' desc;)

```

2. Если прям нужен пооперационный остаток, то следует сортировать таблицу не по id, а по [полю порядок "n"](https://ru.docs.totum.online/tables#table-settings-with_order_field)

PS Архитектурно, таблицу движения денег лучше делать одну и считать остаток в моменте, а не складывать списки с транзакциями

**Reply by gorbarov** [reply]
Я в итоге так и сделал 

```
= : math`$inc-$out`
inc: listSum(list: $incList)
incList: selectList(table: 'income'; field: 'summa_fakt'; where: 'data_fakt' <= #data ; where: 'summa_fakt'>0)

out: listSum(list: $outList)
outList: selectList(table: 'outcome'; field: 'summa_fakt'; where: 'data_fakt' <= #data ; where: 'summa_fakt'>0)
```

В принципе всё работает, да хотелось бы видеть остаток пооперационно. Хотя это не критично, но пооперационно выглядит повеселее ))

**Reply by gorbarov** [reply]
А есть какая-то возможность временную таблицу сбросить так, чтобы ID не двигались, тогда моя модель расчёта начинает работать правильно ))

**Reply by gorbarov** [reply]
По поводу архитектурно.

Вопрос вот в чём, эту таблицу используют для планирования и там постоянно что-то меняется в исходных таблицах, я так прикинул, что это куча геммора с пересчётом постоянным.

Или как вы это видите?

**Reply by erpic-ru** [reply]
Я бы шел от задач. Но прежде, стоит в них разобраться и тут явно недостаточно вводных.

1. Отчет о ДДС (cash-flow) - это факт
2. Бюджет ДДС - это план

Если все это нужно сводить в одном месте, то это опять же можно реализовать в одной простой таблице с признаком проведения true/false и фильтром по нему
Для true рассчитывать баланс и сохранять его. При корректном учете значение пересчитывать не нужно.
Для плана, если проводок не очень много – показывать планируемый баланс через форматирование

**Comment by poluninatm** [comment, marked as answer]
Открыть на ее месте новую через linktotable target: “self”

В качестве hash указать результат функции linktodatatable с параметром
hide:true, куда вы передадите данные для нового расчета.


Пт, 28 апр. 2023 г. в 09:09, gorbarov ***@***.***>:

**Reply by gorbarov** [reply]
Я был бы рад чуть большему количеству деталей )))

**Reply by gorbarov** [reply]
Я правильно понимаю, что надо передать только дату, т.к. на её основе пересчитывается всё остальное.

Попробовал так, не получилось. Вернее таблица обновляется, но поле Дата от, тоже возвращается в исходное значение, заданное кодом.

```
=: linkToTable(table: $#ntn; hash: $hash; target: "self";)
		hash: linkToDataTable(table: $#ntn; hide: true; data: json`[{"h_date_from": #h_date_from}]`)
```

**Reply by ttm-alexeypolunin** [reply]
```
=: linkToTable(table: $#ntn; hash: $hash; target: "self";)
		hash: linkToDataTable(table: $#ntn; params: json`{"h_date_from": #h_date_from}`; hide: true)
```

**Reply by ttm-alexeypolunin** [reply]
Если необязательно иметь id с 1 то: https://github.com/totumonline/totum-ru-issues-and-discussions/discussions/358#discussioncomment-5751721

**Comment by poluninatm** [comment]
Вы все верно поняли!
Только параметр data берет row. Нужно убрать квадратные скобки из json.

Пт, 28 апр. 2023 г. в 09:33, gorbarov ***@***.***>:

**Reply by gorbarov** [reply]
Без квадратных скобок вообще не работает, пустую ошибку выдаёт.

![image](https://user-images.githubusercontent.com/388620/235073447-32193890-70b6-4d83-b2f9-04d474c8c6ce.png)

**Comment by poluninatm** [comment]
Все временные таблицы, что могут нагенерировать несколько человек через веб
таким образом - несущественно. Не думаю, что лишние несколько мегабайт базы
напрягут сервер.

Пт, 28 апр. 2023 г. в 09:39, gorbarov ***@***.***>:

**Comment by poluninatm** [comment]
Нужно смотреть ваш код. Как вы это оформили?

Пт, 28 апр. 2023 г. в 09:41, gorbarov ***@***.***>:

**Reply by gorbarov** [reply]
Сделал кнопку в ней код

![image](https://user-images.githubusercontent.com/388620/235075523-a380bef2-1804-4df5-b5fe-7f09d3f8fc57.png)

**Reply by ttm-alexeypolunin** [reply]
https://github.com/totumonline/totum-ru-issues-and-discussions/discussions/358#discussioncomment-5753888

**Comment by poluninatm** [comment]
Выглядит хорошо, ошибки не вижу. Через пару часов на компе проверю. Очень
интересно!!!

Пт, 28 апр. 2023 г. в 09:51, gorbarov ***@***.***>:

**Comment by ttm-alexeypolunin** [comment]
Хм. Я в таких случаях при изменении в h_insert прописывал удаление всех строк. После этого insert автоматически заполнялся в том порядке в котором все лежит в h_insert. Это был дедовский способ поддерживать сортировку при изменении.

**Reply by gorbarov** [reply]
Круто, а я как это реализовать? Потому что я пробовал с удалением всех строк, но куска кода не осталось... да он и не работал )

**Reply by gorbarov** [reply]
@ttm-alexeypolunin можете уточнить как все строки грохнуть, потому что у меня не получилось это провернуть

**Reply by poluninatm** [reply]
`delete: deleteList(table: 'your_table_name'; where: 'id' > 0)`

**Reply by gorbarov** [reply]
Прошу прощения, я неправильно вопрос задал. В какой момент вызвать deleteList при описанной схеме.
Я пробовал на изменение h_insert_date и h_insert, тогда таблица пустой остаётся.

Либо надо это делать через while ?

**Reply by ttm-alexeypolunin** [reply]
Полез посмотреть как это было сделано и обнаружил, что конечно надо рекалкулейтить таблицу после делита:

```
a1= : deleteList(table: $#ntn; where: 'id' != 0)

a2=: reCalculate(table: $#ntn)
```

**Reply by gorbarov** [reply]
Спасибо, этого и не хватало как раз! Потому что переоткрывание таблицы выглядит не так красиво.
